### PR TITLE
retry ConnectionNotEstablished errors

### PR DESCRIPTION
**Pull request by ecwall**
I tried converting the monkey patch from this support ticket. Is this a proper application?

https://github.com/cockroachlabs/support/issues/2155#issuecomment-1475427677

### CODE REVIEW

File: lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb [hunk 1]
```diff
@@ -10,6 +10,20 @@ module TransactionManagerMonkeyPatch
         # exceeded.
         def within_new_transaction(isolation: nil, joinable: true, attempts: 0)
           super(isolation: isolation, joinable: joinable)
+        rescue ActiveRecord::ConnectionNotEstablished => e
+          raise unless retryable? error
+          raise if attempts >= @connection.max_transaction_retries
+
+          attempts += 1
+          sleep_seconds = (2 ** attempts + rand) / 10
```
Comment (chengxiong-ruan) [inline lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb:18]: let's make a method for this `sleep_seconds` and `sleep`, it's needed by the existing error handling below, maybe something like `sleep_rand_seconds`?

File: lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb [hunk 2]
```diff
@@ -10,21 +10,44 @@ module TransactionManagerMonkeyPatch
...
         end
+
+        def serialization_error?(error)
+          errors = [error]
+          errors << error.cause if error.cause
+          errors.any? {|e| e.is_a? PG::TRSerializationFailure }
+        end
+
+        def sleep_rand_seconds()
+          sleep_seconds = (2 ** attempts + rand) / 10
```
Comment (chengxiong-ruan) [inline lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb:48]: `attempts` should be a parameter of the method.

Comment (ecwall) [reply lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb:48]: Fixed.

File: lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb (hunk 1 above)
Comment (ecwall) [reply lib/active_record/connection_adapters/cockroachdb/transaction_manager.rb]: Added.

**Comment by ecwall** [comment]
bors r=chengxiong-ruan

**Pull request merged**